%\VignetteIndexEntry{Analyzing RNA-seq data with the "DESeq2" package}
%\VignettePackage{DESeq2}
%\VignetteEngine{knitr::knitr}

% To compile this document
% library('knitr'); rm(list=ls()); knit('DESeq2.Rnw')

\documentclass[a4paper,11pt]{article}
\usepackage[top=2cm,bottom=2.5cm,left=2cm,right=2cm]{geometry}

\author{M. Evers$^{1}$\\[1em]
\small{$^{1}$ The John Curtin School of Medical Research,}\\\small{The Australian National University, Canberra, Australia}}

\title{Visualisation and functional analysis of single-nucleotide modifications in mRNAs using the RNAModR package}

\begin{document}
\maketitle

\section{Introduction}
Following is a typical workflow for visualising and analysing single-nucleotide modifications:
\begin{enumerate}
  \item Construction of a custom transcriptome.
  \item Mapping of genome alignment-based single-nucleotide modifications to transcript coordinates.
  \item Visualisation of basic metrics, e.g. distribution of sites across different transcript sections.
  \item Enrichment analyses of single-nucleotide modifications.
\end{enumerate}
The RNAModR library is loaded following
<<>>=
library("RNAModR")
@

\section{Reference transcriptome and transcript annotation}
RNAModR analyses the distribution of RNA modifications across mRNA transcripts and therefore requires suitable mRNA transcriptome and transcript annotation data. When RNAModR is run for the first time, a custom transcriptome is constructed based on UCSC RefSeq gene annotations and associated R data are stored locally in a single binary R file. This process can take up to 10 minutes (depending on the computer hardware), and needs to be run only once as part of a new data analysis. Subsequent re-analyses check for existing suitable transriptome data, and load the required data as part of the different anlysis routines.\par

To start our example of an analysis of the m$^{6}$A sample data \cite{Linder}, we construct the necessary transcriptome and annotation data based on the reference genome assembly version GRCh38/hg38 using
<<>>=
BuildTx("hg38");
@

We can plot the length distribution of the different transcript sections using
<<fig.width=8,fig.height=4>>=
LoadRefTx("hg38");
PlotTxSecLength(txBySec, filter = c("5'UTR", "CDS", "3'UTR", "CDS_exon", "Intron"));
@

\section{Mapping RNA modifications to transcript coordinates}
Next we load a BED file containing genome locations of m$^{6}$A modifications from \cite{?}:
<<>>=
bedFile <- system.file("extdata",
                       "miCLIP_m6A_Linder2015_hg38.bed",
                       package = "RNAModR");
sites <- ReadBED(bedFile);
class(sites);
length(sites);
@

We map genome coordinates of the \Sexpr{length(sites)} sites to the transcriptome:
<<results='hide'>>=
posSites <- SmartMap(sites, id = "m6A", refGenome = "hg38");
@
The return object is an RNAModR-specific object of type txLoc, that contains a list of sites with positions within any of the transcript sections (promoter, 5'UTR, CDS, 3'UTR, intron).

We can access a txLoc to get a brief summary of the mapped sites:
<<>>=
info(posSites);
@

We can also show the distribution of sites across different transcript sections in a piechart:
<<>>=
PlotSectionDistribution(posSites, filter = c("5'UTR", "CDS", "3'UTR"));
@ 
\end{document}
